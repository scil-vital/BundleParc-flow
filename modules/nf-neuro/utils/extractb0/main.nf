process UTILS_EXTRACTB0 {
    tag "$meta.id"
    label 'process_single'

    container "scilus/scilpy:2.2.0_cpu"

    input:
    tuple val(meta), path(dwi), path(bval), path(bvec)

    output:
    tuple val(meta), path("*_b0*.nii.gz"), emit: b0
    path "versions.yml"                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extraction_strategy = task.ext.b0_extraction_strategy ? "--$task.ext.b0_extraction_strategy" : "--mean"
    def b0_threshold = task.ext.b0_threshold ? "--b0_threshold $task.ext.b0_threshold" : ""
    def output_series = task.ext.output_series ? "" : "--single-image"
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_dwi_extract_b0 $dwi $bval $bvec ${prefix}_b0.nii.gz \
        $output_series $extraction_strategy $b0_threshold --skip_b0_check

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_dwi_extract_b0 -h

    touch ${prefix}_b0.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
