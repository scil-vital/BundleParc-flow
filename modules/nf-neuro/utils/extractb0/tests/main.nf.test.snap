{
    "utils - extractb0 - mean": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:15:01.126608471"
    },
    "utils - extractb0 - all4d": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:14:08.955750512"
    },
    "utils - extractb0 - stub-run": {
        "content": [
            [
                "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:15:17.980773861"
    },
    "utils - extractb0": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:13:59.374107972"
    },
    "utils - extractb0 - clustermean": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,feb9010e5b72fc6f0c814f9dd54cb710"
                    ]
                ],
                "1": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,feb9010e5b72fc6f0c814f9dd54cb710"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:14:46.185537537"
    },
    "utils - extractb0 - clusterfirst": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,27e4aa847d60b2154d8865d19ba3e06e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,27e4aa847d60b2154d8865d19ba3e06e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:14:32.853873825"
    },
    "utils - extractb0 - allseries": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,46681cb0bd068f8da81010f9e41d10cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-23T01:14:20.577355867"
    }
}